src/java.base/share/classes/java/util/ImmutableCollections.java

```diff
@@ -168,6 +168,8 @@ private ImmutableCollections() { }
     static <E> List<E> listCopy(Collection<? extends E> coll) {
         if (coll instanceof List12 || (coll instanceof ListN && ! ((ListN<?>)coll).allowNulls)) {
```

Comment:
Maybe replace the cast with an instanceof pattern here?

Reply:
@stuart-marks Sorry, missed this notification. I initially had the same idea, but decided against it because it forces me to suppress "rawtypes" since `coll instanceof ListN` is not considered to be a rawtype, but `coll instanceof ListN c` is. And currently it won't allow for `coll instanceof ListN<E> c`...

Reply:
`coll instanceof ListN<?> list` should work.

Reply:
@forax @stuart-marks Yeah, that works. It's unfortunate that it's not possible to match on the actual (generic) type, as then both sides of the || could use type unification to avoid having to do the cast on the return value as well. Submitted a commit to switch to the wildcard version.

Reply:
Avoiding the cast to `List<E>` requires two things:
- we need a way to say at declaration that a parameter type E is covariant, this is not true for List but this is true for ImmutableList (so true for List2 and ListN).
- we need a way to do an union between two types when doing pattern matching .

This former is still in discussion as a feature of Valhalla (it's a part of Dan Smith Phd), the later is still in discussion as a feature of Amber, so maybe in the future this will be possible is the star aligned.